Enable WPACKET to write to the end of the buffer first

mattcaswell opened the pull request:
This adds the ability to WPACKET to write to the end of the buffer first and fill the buffer backwards. It also adds the ability to support DER style length encoding.

levitte commented on test/wpackettest.c:
```diff
@@ -349,6 +353,43 @@ static int test_WPACKET_memcpy(void)
+            || !TEST_true(WPACKET_get_total_written(&pkt, &written))
+            || !TEST_mem_eq(WPACKET_get_curr(&pkt), written, simpleder,
+                            sizeof(simpleder)))
+        return cleanup(&pkt);
```
Would you mind writing in a comment what ASN.1 structure you're writing here?  I do not understand what you're doing here from a structural point of view, or how to draw actual examples from this.

levitte commented:
> WPACKET could really do with something in doc/internal/man3...

Frankly, that's putting it lightly.  It *must* be documented.

levitte commented on crypto/packet.c:
```diff
@@ -232,6 +255,19 @@ static int wpacket_intern_close(WPACKET *pkt, WPACKET_SUB *sub, int doclose)
+            numlenbytes |= 0x80;
+            if (!WPACKET_put_bytes_u8(pkt, numlenbytes))
+                return 0;
+        }
```
So, uhmmmm, do I understand this right, that I need to make a sub-packet for *every* *damn* *item* in an ASN.1 structure?  Seriously?

mattcaswell replied:
I expect there to be DER helper methods for the different DER types. This is just the underlying WPACKET support.

levitte commented on crypto/packet.c:
```diff
@@ -65,15 +65,21 @@ int WPACKET_reserve_bytes(WPACKET *pkt, size_t len, unsigned char **allocbytes)
-    if (allocbytes != NULL)
+    if (allocbytes != NULL) {
         *allocbytes = WPACKET_get_curr(pkt);
+        if (pkt->endfirst)
```
Should be `if (pkt->endfirst && *allocbytes != NULL)`

mattcaswell replied:
Yes - I actually had this fix in my other temporary branch that I shared with you - but it hadn't made it back into this PR

levitte commented:
I haven't looked too closely, so it's possible that only the DER support stuff needs testing with no buffer (see https://github.com/openssl/openssl/pull/11462#discussion_r402534397), but it's also possible that you need to test that more thoroughly in a more general sense.  I suggest you take an extra look.

mattcaswell commented:
Fixup pushed addressing the feedback above. It also adds a `WPACKET_init_null_der` function - which does the same as `WPACKET_init_null` but calculates lengths properly for DER based length encoding.

The NULL packet encoding needs a test, so I've put this into WIP until I've written that - hopefully tomrrow.

levitte commented on crypto/packet.c:
```diff
@@ -232,6 +265,19 @@ static int wpacket_intern_close(WPACKET *pkt, WPACKET_SUB *sub, int doclose)
+        }
     }
 
     if (doclose) {
```
Ok, `packlen` is being written to the buffer, fine, but then what?  At least in DER terms, it needs to be accumulated in the parent packet, otherwise we end up with too short lengths all over the place.  Remember that the length of a SEQUENCE, SET or CHOICE is the sum of the lengths of all its elements, including their tags and lengths...

levitte replied:
Never mind this, I was looking at the wrong byte...

levitte commented:
I've used this merged with the latest #11450, and it now works flawlessly.

mattcaswell commented:
Now updated with some additional test code to cover the `WPACKET_init_null_der` function. Taken out of WIP. Please take a look.

levitte commented:
Travis failure is the usual timeout...  why is this still happening?

richsalz commented:
> Travis failure is the usual timeout... why is this still happening?

Because *one hour* is still not enough time.

This build should be removed.

levitte commented:
> This build should be removed.

PR welcome :wink:

richsalz commented:
I would do remove the build if I could, but I don't know enough about travis configuration matrices, sorry.

mattcaswell commented:
> This build should be removed.

I suggest instead a stripped back build that runs a subset of the tests.

mattcaswell commented:
Pushed. Thanks.
